lib/Sema/CSBindings.cpp

```diff
@@ -981,7 +981,47 @@ bool BindingSet::isViable(PotentialBinding &binding) {
 }
 
 bool BindingSet::favoredOverDisjunction(Constraint *disjunction) const {
-  if (isHole() || isDelayed())
+  if (isHole())
+    return false;
+
+  if (llvm::any_of(Bindings, [&](const PotentialBinding &binding) {
+        if (binding.Kind == AllowedBindingKind::Supertypes)
+          return false;
+
+        auto type = binding.BindingType;
+
+        if (CS.shouldAttemptFixes())
+          return false;
+
+        if (type->isAnyHashable() || type->isDouble() || type->isCGFloatType())
+          return false;
+
+        {
+          PointerTypeKind pointerKind;
+          if (type->getAnyPointerElementType(pointerKind)) {
+            switch (pointerKind) {
+            case PTK_UnsafeRawPointer:
+            case PTK_UnsafeMutableRawPointer:
+              return false;
+
+            default:
+              break;
+            }
+          }
+        }
+
+        return type->is<StructType>() || type->is<EnumType>();
+      })) {
+    // Result type of subscript could be l-value so we can't bind it early.
+    if (!TypeVar->getImpl().isSubscriptResultType() &&
```

Comment:
Ah got it, thanks for the detailed explanation! 

> Subscript is special because that's the only "call" that is allowed to return l-value result

That would be because of `_modify`/`_yeild` coroutine things allowed only in subscripts I would guess  

> I think in the future it might be interesting to experiment with adding l-value type to the binding set or allow binding producer to attempt it for subscript results...

👍